src/backend/distributed/transaction/backend_data.c

```diff
@@ -1278,3 +1293,87 @@ DecrementExternalClientBackendCounter(void)
 {
 	pg_atomic_sub_fetch_u32(&backendManagementShmemData->externalClientBackendCounter, 1);
 }
+
+
+/*
+ * ResetCitusBackendType resets the backend type cache.
+ */
+void
+ResetCitusBackendType(void)
+{
+	CurrentBackendType = CHECK_APPLICATION_NAME;
+}
+
+
+/*
+ * IsRebalancerInitiatedBackend returns true if we are in a backend that citus
+ * rebalancer initiated.
+ */
+bool
+IsRebalancerInternalBackend(void)
+{
+	if (CurrentBackendType == CHECK_APPLICATION_NAME)
+	{
+		DetermineCitusBackendType();
+	}
+
+	return CurrentBackendType == CITUS_REBALANCER_BACKEND;
+}
+
+
+/*
+ * IsCitusInitiatedRemoteBackend returns true if we are in a backend that citus
+ * initiated via remote connection.
+ */
+bool
+IsCitusInternalBackend(void)
+{
+	if (CurrentBackendType == CHECK_APPLICATION_NAME)
+	{
+		DetermineCitusBackendType();
+	}
+
+	return CurrentBackendType == CITUS_INTERNAL_BACKEND;
+}
+
+
+/*
+ * IsCitusRunCommandBackend returns true if we are in a backend that one of
+ * the run_command_on_* functions initiated.
+ */
+bool
+IsCitusRunCommandBackend(void)
+{
+	if (CurrentBackendType == CHECK_APPLICATION_NAME)
+	{
+		DetermineCitusBackendType();
+	}
+
+	return CurrentBackendType == CITUS_RUN_COMMAND_BACKEND;
+}
+
+
+/*
+ * DetermineCitusBackendType determines the type of backend based on the application_name.
+ */
+static void
+DetermineCitusBackendType(void)
```

Comment:
this PR actually makes changing gpid as well if we want in the future. We can basically call AssignGlobalPID somewhere after DetermineCitusBackendType.

But, we don't need it right now, just keeping an eye on it